cpp/src/community/k_truss_impl.cuh

```diff
@@ -0,0 +1,929 @@
+#include <thrust/scatter.h>
+#include <thrust/sort.h>
+#include <thrust/transform.h>
+#include <thrust/tuple.h>
+
+#include <prims/edge_bucket.cuh>
+#include <prims/extract_transform_e.cuh>
+#include <prims/fill_edge_property.cuh>
+#include <prims/reduce_op.cuh>
+#include <prims/transform_e.cuh>
+#include <prims/transform_reduce_dst_nbr_intersection_of_e_endpoints_by_v.cuh>
+#include <prims/update_edge_src_dst_property.cuh>
+
+namespace cugraph {
+
+template <typename vertex_t, typename edge_t, typename EdgeIterator>
+struct unroll_edge {
+  edge_t num_valid_edges{};
+  raft::device_span<edge_t> num_triangles{};  // FIXME: invalid type for unsigned integers
+  EdgeIterator edge_unrolled{};
+  EdgeIterator transposed_valid_edge_first{};
+  EdgeIterator transposed_valid_edge_last{};
+  EdgeIterator transposed_invalid_edge_last{};
+
+  __device__ thrust::tuple<vertex_t, vertex_t> operator()(edge_t i) const
+  {
+    // edges are sorted with destination as key so reverse the edge when looking it
+    auto pair = thrust::make_tuple(thrust::get<1>(*(edge_unrolled + i)),
+                                   thrust::get<0>(*(edge_unrolled + i)));
+    // Find its position in either partition of the transposed edgelist
+    // An edge can be in found in either of the two partitions (valid or invalid)
+
+    auto itr = thrust::lower_bound(
+      thrust::seq, transposed_valid_edge_last, transposed_invalid_edge_last, pair);
+    auto idx = thrust::distance(transposed_valid_edge_last, itr) + num_valid_edges;
+
+    if (*itr != pair) {
+      // The edge must be in the first boundary
+      itr = thrust::lower_bound(
+        thrust::seq, transposed_valid_edge_first, transposed_valid_edge_last, pair);
+      idx = thrust::distance(transposed_valid_edge_first, itr);
+    }
+
+    assert(*itr == pair);
+    cuda::atomic_ref<edge_t, cuda::thread_scope_device> atomic_counter(num_triangles[idx]);
+    auto r = atomic_counter.fetch_sub(edge_t{1}, cuda::std::memory_order_relaxed);
+  }
+};
+
+template <typename vertex_t, typename edge_t>
+struct unroll_edge_invalid {
+  edge_t num_valid_edges{};
+  raft::device_span<edge_t> num_triangles{};  // FIXME: invalid type for unsigned integers
+  raft::device_span<edge_t> num_triangles_invalid{};
+
+  __device__ thrust::tuple<vertex_t, vertex_t> operator()(edge_t idx) const
+  {
+    cuda::atomic_ref<edge_t, cuda::thread_scope_device> atomic_counter(
+      num_triangles[idx + num_valid_edges]);
+    auto r =
+      atomic_counter.fetch_sub(edge_t{num_triangles_invalid[idx]}, cuda::std::memory_order_relaxed);
+  }
+};
+
+template <typename vertex_t, typename edge_t>
+rmm::device_uvector<vertex_t> prefix_sum_valid_and_invalid_edges(
+  // The edgelist is segmented into 2 partitions (valid and invalid edges)
+  // and edges to be unrolled can be either in the valid or the invalid edge
+  // partition.
+  raft::handle_t const& handle,
+  edge_t num_edges,
+  edge_t num_invalid_edges,
+  vertex_t* invalid_dst,
```

Comment:
raft::device_span<vertex_t const>, mixing raw pointers and device_spans is confusing.